src/core/ext/gcp/metadata_query.cc

```diff
@@ -0,0 +1,131 @@
+    "/computeMetadata/v1/instance/id";
+constexpr const char MetadataQuery::kIPv6Attribute[] =
+    "/computeMetadata/v1/instance/network-interfaces/0/ipv6s";
+
+MetadataQuery::MetadataQuery(
+    std::string attribute, grpc_polling_entity* pollent,
+    absl::AnyInvocable<void(std::string /* attribute */,
+                            absl::StatusOr<std::string> /* result */)>
+        callback,
+    Duration timeout)
+    : InternallyRefCounted<MetadataQuery>(nullptr, 2),
+      attribute_(std::move(attribute)),
+      callback_(std::move(callback)) {
+  GRPC_CLOSURE_INIT(&on_done_, OnDone, this, nullptr);
+  auto uri = URI::Create("http", "metadata.google.internal.", attribute_,
+                         {} /* query params */, "" /* fragment */);
+  GPR_ASSERT(uri.ok());  // params are hardcoded
+  grpc_http_request request;
+  memset(&request, 0, sizeof(grpc_http_request));
+  grpc_http_header header = {const_cast<char*>("Metadata-Flavor"),
+                             const_cast<char*>("Google")};
+  request.hdr_count = 1;
+  request.hdrs = &header;
+  http_request_ = HttpRequest::Get(
+      std::move(*uri), nullptr /* channel args */, pollent, &request,
+      Timestamp::Now() + timeout, &on_done_, &response_,
+      RefCountedPtr<grpc_channel_credentials>(
+          grpc_insecure_credentials_create()));
+  http_request_->Start();
+}
+
+MetadataQuery::~MetadataQuery() { grpc_http_response_destroy(&response_); }
+
+void MetadataQuery::Orphan() {
+  http_request_.reset();
+  Unref();
+}
+
+void MetadataQuery::OnDone(void* arg, absl::Status error) {
+  auto* self = static_cast<MetadataQuery*>(arg);
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_metadata_query_trace)) {
+    gpr_log(GPR_INFO, "MetadataServer Query for %s: HTTP status: %d, error: %s",
+            self->attribute_.c_str(), self->response_.status,
+            StatusToString(error).c_str());
+  }
+  absl::StatusOr<std::string> result;
+  if (!error.ok()) {
+    result = absl::UnavailableError(absl::StrFormat(
+        "MetadataServer Query failed for %s: %s", self->attribute_.c_str(),
+        StatusToString(error).c_str()));
+  } else if (self->response_.status != 200) {
+    result = absl::UnavailableError(absl::StrFormat(
+        "MetadataServer Query received non-200 status for %s: %s",
+        self->attribute_.c_str(), StatusToString(error).c_str()));
+  } else if (self->attribute_ == kZoneAttribute) {
+    absl::string_view body(self->response_.body, self->response_.body_length);
+    size_t pos = body.find_last_of('/');
+    if (pos == body.npos) {
+      result = absl::UnavailableError(
+          absl::StrFormat("MetadataServer Could not parse zone: %s",
+                          std::string(body).c_str()));
+      if (GRPC_TRACE_FLAG_ENABLED(grpc_metadata_query_trace)) {
+        gpr_log(GPR_INFO, "%s", result.status().ToString().c_str());
+      }
+    } else {
+      result = std::string(body.substr(pos + 1));
+    }
+  } else {
+    result = self->response_.body;
```

Comment:
Is the body guaranteed to be nul-terminated and to contain no nul characters before the end?  If not, this isn't safe.  I think instead it should be:

```
result = std::string(self->response_.body, self->response_.body_length);
```

Reply:
Done. Good catch! Thanks!